sympy/polys/polytools.py

```diff
@@ -3816,6 +3816,72 @@ def nth_power_roots_poly(f, n):
 
         return r.replace(t, x)
 
+    def which_roots(f, candidates, real=True):
+        """
+        Given a superset of roots of f, finds which ones are
+        roots of f. Note this won't work properly if candidates is
+        not a superset of the roots of f. The list returned is the
+        same length as the number of roots of f, counted with
+        multiplicity, and preserves the order of the original candidates list.
+
+        Examples
+        ========
+
+        >>> from sympy import Poly, I
+        >>> from sympy.abc import x
+
+        >>> f = Poly(x**4 - 1)
+
+        >>> f.which_roots([-1, 1, 0, -2, 2], real=True)
+        [-1, 1]
+        >>> f.which_roots([-1, 1, -I, I, 0], real=False)
+        [-1, 1, -I, I]
+        >>> f.which_roots([2, 5], real=True) # expectedly wrong
+        [2, 5]
```

Comment:
Rather than showing this example I think it might be clearer to show a demonstration involving `lift`.

Also it would be good to have:
```
See Also
=======

same_root
```
Likewise `same_root` could reference `which_roots`.

Reply:
Sounds good, will do